File: logic/asm/elf/src/bind/section.cpp
#include "section.hpp"
#include "utils.hpp"

bind::Section::Section(const Napi::CallbackInfo &info)
    : Napi::ObjectWrap<Section>(info) {
  bind::detail::count_args(info, 2, 2);
  this->elf =
      *bind::detail::parse_arg_external<std::shared_ptr<ELFIO::elfio>>(info, 0, "shared pointer to an elf file");
  this->section = bind::detail::parse_arg_external<ELFIO::section>(info, 1, "raw pointer to an elf segment");
}

Napi::Value bind::Section::get_index(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_index()));
}

Napi::Value bind::Section::get_name(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::String::New(info.Env(), section->get_name());
}

Napi::Value bind::Section::set_name(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_name(bind::detail::parse_arg_string(info, 0, "string"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_type(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_type()));
}

Napi::Value bind::Section::set_type(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_type(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_flags(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_flags()));
}

Napi::Value bind::Section::set_flags(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_flags(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_info(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_info()));
}

Napi::Value bind::Section::set_info(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_info(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_link(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_link()));
}

Napi::Value bind::Section::set_link(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_link(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_align(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_addr_align()));
}

Napi::Value bind::Section::set_align(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_addr_align(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_address(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_address()));
}

Napi::Value bind::Section::set_address(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  section->set_address(bind::detail::parse_arg_bigint(info, 0, "bigint"));
  return info.Env().Null();
}

Napi::Value bind::Section::get_size(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 0, 0);
  return Napi::BigInt::New(info.Env(), static_cast<uint64_t>(section->get_size()));
}

Napi::Value bind::Section::set_data(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  auto array = bind::detail::parse_arg_uint8array(info, 0, "Uint8Array");
  auto buffer = array.ArrayBuffer();
  section->set_data((const char *) buffer.Data(), buffer.ByteLength());
  return info.Env().Null();
}

Napi::Value bind::Section::append_data(const Napi::CallbackInfo &info) {
  bind::detail::count_args(info, 1, 1);
  auto array = bind::detail::parse_arg_uint8array(info, 0, "Uint8Array");
  auto buffer = array.ArrayBuffer();
  section->append_data((const char *) buffer.Data(), buffer.ByteLength());
  return info.Env().Null();
}

Napi::Function bind::Section::GetClass(Napi::Env env) {
  return bind::Section::DefineClass(env, "Section", {
      Section::InstanceMethod("getIndex", &Section::get_index),
      Section::InstanceMethod("getName", &Section::get_name),
      Section::InstanceMethod("setName", &Section::set_name),
      Section::InstanceMethod("getType", &Section::get_type),
      Section::InstanceMethod("setType", &Section::set_type),
      Section::InstanceMethod("getFlags", &Section::get_flags),
      Section::InstanceMethod("setFlags", &Section::set_flags),
      Section::InstanceMethod("getInfo", &Section::get_info),
      Section::InstanceMethod("setInfo", &Section::set_info),
      Section::InstanceMethod("getLink", &Section::get_link),
      Section::InstanceMethod("setLink", &Section::set_link),
      Section::InstanceMethod("getAlign", &Section::get_align),
      Section::InstanceMethod("setAlign", &Section::set_align),
      Section::InstanceMethod("getAddress", &Section::get_address),
      Section::InstanceMethod("setAddress", &Section::set_address),
      Section::InstanceMethod("getSize", &Section::get_size),
      Section::InstanceMethod("setData", &Section::set_data),
      Section::InstanceMethod("appendData", &Section::append_data),
  });
}

File: logic/asm/elf/src/bind/section.ts
/* eslint-disable camelcase */
export const sh_type = {
  SHT_NULL: 0n,
  SHT_PROGBITS: 1n,
  SHT_SYMTAB: 2n,
  SHT_STRTAB: 3n,
  SHT_RELA: 4n,
  SHT_HASH: 5n,
  SHT_DYNAMIC: 6n,
  SHT_NOTE: 7n,
  SHT_NOBITS: 8n,
  SHT_REL: 9n,
  SHT_SHLIB: 10n,
  SHT_DYNSYM: 11n,
  SHT_LOPROC: 0x70000000n,
  SHT_HIPROC: 0x7fffffffn,
  SHT_LOUSER: 0x80000000n,
  SHT_HIUSER: 0xffffffffn,
} as const;

export const sh_flags = {
  SHF_WRITE: 0x1n,
  SHF_ALLOC: 0x2n,
  SHF_EXECINSTR: 0x4n,
  SHF_MERGE: 0x10n,
  SHF_STRINGS: 0x20n,
  SHF_INFO_LINK: 0x40n,
  SHF_LINK_ORDER: 0x80n,
  SHF_OS_NONCONFORMING: 0x100n,
  SHF_GROUP: 0x200n,
  SHF_TLS: 0x400n,
  SHF_COMPRESSED: 0x800n,
  SHF_GNU_RETAIN: 0x200000n,
  SHF_GNU_MBIND: 0x01000000n,
  SHF_MASKOS: 0x0FF00000n,
  SHF_MIPS_GPREL: 0x10000000n,
  SHF_ORDERED: 0x40000000n,
  SHF_EXCLUDE: 0x80000000n,
  SHF_MASKPROC: 0xF0000000n,
} as const;

export interface SectionHeader {
    name: string
    /** @see {@link sh_type} */
    type: bigint
    /** @see {@link sh_flags} */
    flags: bigint
    addr: bigint
    /** Must be 0 or the index of another section */
    link: bigint
    /** Depends on sh_type, see TIS ELF spec Fig 1-12 */
    info: bigint
    /** Must be power of 2 */
    addr_align: bigint
}

export interface Section {
    getIndex(): bigint

    getName(): string
    setName(name:string): void

    /** @see {@link sh_type} */
    getType(): bigint
    /**
     * @arg type should be from sh_type enum
     * @see {@link sh_type}
     */
    setType(type:bigint): void

    setEntrySize(size:bigint):void

    /** @see {@link sh_flags} */
    getFlags(): bigint
    /**
     * @arg type should be from sh_type enum
     * @see {@link sh_flags}
     */
    setFlags(flags:bigint): void

    getInfo(): bigint
    setInfo(info:bigint): void

    getLink(): bigint
    setLink(link:bigint): void

    getAlign(): bigint
    setAlign(align:bigint): void

    getAddress(): bigint
    setAddress(address: bigint): void

    // Can't set, because it is computed from data
    getSize(): bigint
    // Force override size value.
    setSize(size:bigint):void;
    setData(bytes:Uint8Array): void
    appendData(bytes:Uint8Array): void
    getData(): Uint8Array
}

export const markSymtab = (section: Section, link:bigint, entrySize:bigint) => {
  section.setType(sh_type.SHT_SYMTAB);
  section.setAlign(0n);
  section.setLink(link);
  section.setAlign(0x2n);
  section.setEntrySize(entrySize);
};

export const markStrtab = (section: Section) => {
  section.setType(sh_type.SHT_STRTAB);
  section.setAlign(0n);
};
